Core: On demand store

tmeasday opened the pull request:
### From the migration notes:

Introduces an opt-in feature flag, `features.storyStoreV7`, which loads stories in an "on demand" way (that is when rendered), rather than up front when the Storybook is booted. This way of operating will become the default in 7.0 and will likely be switched to opt-out in that version.

The key benefit of the on demand store is that stories are code-split automatically (in `builder-webpack4` and `builder-webpack5`), which allows for much smaller bundle sizes, faster rendering, and improved general performance via various opt-in Webpack features.

The on-demand store relies on the "story index" data structure which is generated in the server (node) via static code analysis. As such, it has the following limitations:

- Does not work with `storiesOf()`
- Does not work if you used dynamic story names or component titles.

However, the `autoTitle` feature is supported.

#### Behavioral differences

The key behavioral differences of the v7 store are:

- `SET_STORIES` is not emitted on boot up. Instead the manager loads the story index independenly.
- A new event `STORY_PREPARED` is emitted when a story is rendered for the first time, which contains metadata about the story, such as `parameters`.
- All "entire" store APIs such as `extract()` need to be proceeded by an async call to `loadAllCSFFiles()` which fetches all CSF files and processes them.

#### Using the v7 store

To activate the v7 mode set the feature flag in your `.storybook/main.js` config:

```js
module.exports = {
  // ... your existing config
  features: {
    storyStoreV7: true,
  },
};
```

NOTE: `features.storyStoreV7` implies `features.buildStoriesJson` and has the same limitations.

---

This PR is too large to be sensibly reviewed, instead it has been split into several smaller PRs that can be examined separately and changes will be cherry picked back here:

- [ ] https://github.com/storybookjs/storybook/pull/15999
- [ ] https://github.com/storybookjs/storybook/pull/16004
- [ ] https://github.com/storybookjs/storybook/pull/16005
- [ ] https://github.com/storybookjs/storybook/pull/16006
- [ ] https://github.com/storybookjs/storybook/pull/16009
- [ ] https://github.com/storybookjs/storybook/pull/16010
- [ ] https://github.com/storybookjs/storybook/pull/16011
- [ ] https://github.com/storybookjs/storybook/pull/16012
- [ ] https://github.com/storybookjs/storybook/pull/16034

--- 

Current limitations:

 - [ ] HMR is not fully supported in the v7 store, adding/removing stories will not be picked up, you need to reload the page manually.
 - [ ] v7 mode currently does not work with Storyshots
 - [ ] v7 mode currently does not work with Composition
 - [ ] v7 mode currently does not work with Links
 - [ ] No UI is currently show for async story rendering (not necessarily a big problem but could be improved).

These limitations will be addressed in follow up PRs.

gaetanmaisse committed:
Revert "fix(ArgControl): unbox the value to simplify the code"

This reverts commit e42fcf0d3d5d25e1abab021c7c28a160b017387b.

tmeasday committed:
Remove logging

tmeasday committed:
Making progress through the tests

shilman committed:
Merge branch 'on-demand-store' of github.com:storybookjs/storybook into on-demand-store

shilman committed:
Fix undefined story parameters

shilman committed:
Addon-docs: Fix description block

tmeasday committed:
More WebPreview tests

tmeasday committed:
Finished docs=>story tests

tmeasday committed:
Completed changing story tests

tmeasday committed:
Wrote story change during rendering tests

tmeasday committed:
Finished story=>docs tests

tmeasday committed:
Added some initial HMR tests

tmeasday committed:
Added couple of full stack tests

tmeasday committed:
Split out integration test, added rudimentary docs one

tmeasday committed:
Update validateOptions behaviour and add test

shilman committed:
Merge branch 'on-demand-store' of github.com:storybookjs/storybook into on-demand-store

tmeasday committed:
Retain a delta of arg changes on HMR

If you update an arg value, then change the underlying (initial) args, we look at what you've changed, then try and apply it over the new initial args.

tmeasday committed:
Add a test that arg changes are retained as you browse around

tmeasday committed:
Updated globals behaviour on HMR to mirror args

tmeasday committed:
Ensure we don't keep "spurious" changes on implementation updates

Also this ensures weird stuff doesn't happen when we change arg/global names.

tmeasday committed:
Remove tests that have been ported :+1:

tmeasday committed:
Move tests around selection specifier + HMR over

tmeasday committed:
Fix linting / compilation problems

tmeasday committed:
Repair modules/stories tests

tmeasday committed:
Fix WebPreview tests

tmeasday committed:
Fix `enhanceArgTypes`

tmeasday committed:
Fix addArgsHelpers test

shilman committed:
Official-storybook: Remove template strings (temporary)

shilman committed:
Official-storybook: Remove design-system (temporary)

tmeasday committed:
Move new files from `client-api` => `store`

shilman committed:
Official-storybook: Fix main.ts config for on-demand store

tmeasday committed:
Moved stuff to webpreview package

tmeasday committed:
Semi fix/hack client-api/core-client types so they build

tmeasday committed:
Need SB Types

tmeasday committed:
Updated all render function to get build working

tmeasday committed:
Fix up UrlStore + internal types

tmeasday committed:
Fix up `inferControls`

tmeasday committed:
Migrate a lot of types in the on-demand-store

tmeasday committed:
A lot of type changes, not quite building yet

tmeasday committed:
Types + tests working in store

tmeasday committed:
Don't do autotitle any more, get it from stories list

tmeasday committed:
Got addon-controls working

tmeasday committed:
Sorted out inferArgTypes

tmeasday committed:
Cleanup

tmeasday committed:
Emit event on config error

tmeasday committed:
Implemented `ClientApi` shim

tmeasday committed:
Got first version of storiesOf shim in. Not yet running

tmeasday committed:
Update CSF

tmeasday committed:
Close to bootstrapping again

tmeasday committed:
Got things bootstrappign

tmeasday committed:
Got old-style working

tmeasday committed:
Some improvements

tmeasday committed:
Fixed some problems and added some tests for start.ts

tmeasday committed:
Added a test for the combined case

tmeasday committed:
Added `forceReRender` to start

tmeasday committed:
Clean code up a little

tmeasday committed:
Handle HMR in storiesOf

tmeasday committed:
HMR tests for start

tmeasday committed:
Added a new feature `storyStoreV7`

tmeasday committed:
Maintain component order on HMR

tmeasday committed:
Small cleanups

tmeasday committed:
We need to refetch the story list when the import fn changes

tmeasday committed:
Implemented .extract() and better getSetStoriesPayload() on store

tmeasday committed:
Small type fix

tmeasday committed:
Got getStorybook() going, refactor start + client-api

tmeasday committed:
Somehow this got lost

tmeasday committed:
Fix type

tmeasday committed:
Add errors for removed APIs

tmeasday committed:
Update decorators to use a common sanitization

tmeasday committed:
Type finangling

tmeasday committed:
Upgrade CSF again

tmeasday committed:
Clean up some exported APIs

tmeasday committed:
Don't export `ConfigApi` any more -- not sure it could be used directly anyway

tmeasday committed:
Throw from client api exports if using storyStoreV7

tmeasday committed:
Ensure SB globals are always defined

tmeasday committed:
Re-established old addons types

tmeasday committed:
Re-add CSF types to addons

tmeasday committed:
Fixed up ClientApi types

tmeasday committed:
Update types in core-client

tmeasday committed:
Got core-client going too

tmeasday committed:
Cleanup

tmeasday committed:
More cleanup

tmeasday committed:
Update frameworks to use legacy types

tmeasday committed:
Update a bunch of types to get things to 'strap

tmeasday committed:
Added a "sync" init path to allow storyshots to be sycn

tmeasday committed:
Don't watch a channel just yet in SLS

tmeasday committed:
Fixed up getStorybook/raw and made storyshots work

tmeasday committed:
Lock update

tmeasday committed:
Reverse Story/UnboundStory

tmeasday committed:
Update argTypes tests

tmeasday committed:
Fix up angular decorators

tmeasday committed:
Pass component through angular properly (not on parameters any more)

tmeasday committed:
Little improvement to `inferArgTypes`

tmeasday committed:
Repair globals/globalTypes in legacy

tmeasday committed:
Temporary work arounds for a couple of angular issues

tmeasday committed:
Fix vue+svelte shots

tmeasday committed:
Update snapshot (looks like this one was broken before)

tmeasday committed:
Update `yarn.lock`

tmeasday committed:
No need to mock in this test

tmeasday committed:
Move inferControls from the controls addon back into the store

For backwards compatibility reasons, we will undo this commit in 7.0

tmeasday committed:
Move inferArgTypes back into the store

So we don't need to do it in Storyshots anymore.

tmeasday committed:
Fix angular stories by moving component back into return value

tmeasday committed:
Set legacy parameters in context for back-compat.

Also add a `breakingChangesV7` feature to turn of things like this to check future compatibility.

tmeasday committed:
Fix addon-docs tests

tmeasday committed:
Change `Framework` => `TFramework`

tmeasday committed:
Whoops this was supposed to be optional

tmeasday committed:
Use breaking changes in store/preview tests

tmeasday committed:
Fix withActions

tmeasday committed:
Simplified iframe-webpack.config.js thanks to @ndelangen's feedback

tmeasday committed:
Bring wp4 changes over to wp5

tmeasday committed:
Remove incomplete tests for `to-importFn`

I want to revisit `to-importFn` as part of integrating the dynamic stories list server, and so will leave it without tests for now.

tmeasday committed:
Add a note about backcompat

tmeasday committed:
Add template to wp5 builder

tmeasday committed:
Add a test that `.extract()` filters docs-only

tmeasday committed:
Need to pick filenames carefully in configs

tmeasday committed:
Added useStoryPrepared() hook

tmeasday committed:
Fix global handling

tmeasday committed:
Fix stories tests and added some for preparing

tmeasday committed:
Fix ui mockdata

tmeasday committed:
Update `prepareStory` tests

tmeasday committed:
Simplify start tests by turning on breaking v7 changes

tmeasday committed:
Update vue snapshots with backcompat parameters

tmeasday committed:
Ensure user's config is last on the list.

And update snapshots with cosmetic changes to entry points (and ordering HMR to the end, which I think is OK)

ghengeveld commented on addons/a11y/package.json:
```diff
@@ -48,10 +48,10 @@
     "@storybook/client-logger": "6.4.0-alpha.19",
     "@storybook/components": "6.4.0-alpha.19",
     "@storybook/core-events": "6.4.0-alpha.19",
+    "@storybook/csf": "0.0.2--canary.a1972cb.0",
```
Shouldn't this be `@componentdriven/csf`?

ghengeveld commented on addons/controls/package.json:
```diff
@@ -47,11 +47,14 @@
+    "@storybook/store": "6.4.0-alpha.19",
     "@storybook/theming": "6.4.0-alpha.19",
     "core-js": "^3.8.2",
+    "lodash": "^4.17.20",
```
I don't see Lodash being used anywhere.

tmeasday committed:
Updated `caniuse`

tmeasday committed:
Drop unused linting comments

tmeasday committed:
Add `StoryStore.getStoriesJsonData()`

tmeasday committed:
Remove a lot of TODOs, clean up

tmeasday committed:
Don't use title exports in non v7 store

tmeasday committed:
Didn't mean to commit this

tmeasday committed:
Drop a couple FIXMEs that were fixed

tmeasday committed:
Deal with multiple `storiesOf` in the same file

tmeasday committed:
Ensure `render` is used in legacy mode

tmeasday committed:
Move sorting types back to addons

tmeasday committed:
Add sorting to back-compat store

tmeasday committed:
Update extract to wait for init

tmeasday committed:
Undo extract change and fix up web-components-kitchen-sink

tmeasday committed:
Fix incorrect link

tmeasday committed:
Update CSF version to pass Args through to stories

tmeasday committed:
Don't put a `StoryFn` onto `render`

As it is supposed to be an `ArgsStoryFn`

tmeasday committed:
Various fixes

tmeasday committed:
Ensure we deal with cleaning up docs only too

tmeasday committed:
Don't generate stories.json if we don't have to

tmeasday committed:
Re-add client-api to builder-webpack4

tmeasday committed:
Ensure the user (i.e. storyshots) fileName wins

tmeasday committed:
No idea why I need to update these angular snapshots

tmeasday committed:
Add store to builder-webpackX deps

tmeasday committed:
Add regenerator to store and web-preview

tmeasday committed:
Fix issue with args memory over HMR applying in the wrong place

tmeasday committed:
Ensure we reset args *before* re-rendering

tmeasday committed:
Remove old comments

tmeasday committed:
Use object freeze to ensure stories are not editable

tmeasday committed:
Switch `createGate` to a hooks-like syntax

tmeasday committed:
Ensure we handle it if loaders or play throw

tmeasday committed:
Fix old use of "Meta" (=> "Annotation")

tmeasday committed:
Refactor to pass sync as an arg rather than using a separate method

Using TS overloads to deal with the type differences. It's a little awkward but probably avoids code duplication a little.

tmeasday committed:
Rename GlobalAnnotations => ProjectAnnotations

tmeasday committed:
Rename `WebPreview` to `PreviewWeb`

tmeasday committed:
Rename `StoriesList` => `StoryIndex`

tmeasday committed:
Small fixes to get boostrapping

tmeasday committed:
Split ClientApi facade into its own class

tmeasday committed:
Add a "back-compat" API for `__STORYBOOK_CLIENT_API`

(Although it isn't backwards compatible, as it requires you call `cacheAllCSFFiles()` first)

tmeasday committed:
Cleanup

tmeasday committed:
Deal with initial preview error

tmeasday committed:
Add async-ness to Story+ArgsTable blocks

tmeasday committed:
Update CSF version (projectAnnotations)

tmeasday committed:
Fix bootstrap

tmeasday committed:
Deepscan cleanups

tmeasday committed:
Update snapshots due to WP => PW change

tmeasday committed:
Remove unusued function arg (deepscan)

tmeasday committed:
Merge remote-tracking branch 'origin/next' into on-demand-store

tmeasday committed:
Get it bootstrapping again

tmeasday committed:
Update CSF

tmeasday committed:
Tweak `ArgTypes`

tmeasday committed:
Fix handlebars dep

tmeasday committed:
Improve types flushed out by CSF changes

tmeasday committed:
Fix a bunch of types

tmeasday committed:
Remove autotitle story from angular

tmeasday committed:
Fix broken `core-server` tests

tmeasday committed:
Update two angular snapshots

tmeasday committed:
Fix types in a couple of angular tests

tmeasday committed:
Remove unused import

tmeasday committed:
Update csf dep in csf-tools

tmeasday committed:
Switch order of args / argTypes enhancers

tmeasday committed:
Add getter for `_storyStore` (back compat)

tmeasday committed:
Fix mdx-in-story story

tmeasday committed:
Fix argtypes stories

tmeasday committed:
Fix ember

tmeasday committed:
Remove autotitle story from angular SB / fix other

tmeasday committed:
Ensure we run argTypes enhancers before pulling defaultValues

tmeasday committed:
Wait a tick before apply arg changes and other re-renders.

This allows users to update args or re-render the story from *inside* decorators or story functions, for better or worse.

tmeasday committed:
Fix globals

tmeasday committed:
Abort controller fallback for IE

tmeasday committed:
Update CSF to allow `symbol`

tmeasday committed:
Ensure we teardown when going to missing properly

tmeasday committed:
Add a note

tmeasday committed:
Don't emit `STORY_PREPARED` in v6 mode.

It was causing problems when there was nothing in the store on the manager side.

tmeasday committed:
Don't allow setting args to `undefined`

tmeasday committed:
Ignore v<3 storyIndexes

tmeasday committed:
Ensure we send `args` with `SET_STORIES`

tmeasday committed:
Fix storyshots

tmeasday committed:
Revert "Wait a tick before apply arg changes and other re-renders."

This reverts commit 23254408d16d3357b72a05903a85e76b1639e663.

tmeasday committed:
Get rid of setTimeout(0) and instead just call `renderToDOM` immediately

The framework can do the queuing

tmeasday committed:
Added a bunch of comments and some READMEs

tmeasday committed:
Arg fix `lib/api`

tmeasday committed:
Fix exported type from `.raw()`

tmeasday committed:
Add migration notes about new features

tmeasday committed:
Merge pull request #15987 from storybookjs/on-demand-store-sync-arg

Refactor to pass sync as an arg rather than using a separate method

shilman committed:
Temporarily restore glob-base to fix importFn

shilman committed:
Recompute stories.json on each request in dev-mode

tmeasday committed:
Add warning if find old `stories.json`

https://github.com/storybookjs/storybook/pull/16004/files/b05020ce5c1ca9f12cc5c0f7bd4e2aa8d20d7102#r705834795

tmeasday committed:
Apply suggestions from code review

tmeasday committed:
Add presets for all frameworks

tmeasday committed:
Somehow missed this

tmeasday committed:
Somehow missed this too

tmeasday committed:
Update most <any> to nothing which defaults to <AnyFramework>

tmeasday committed:
Update CSF for `AnyFramework`

tmeasday committed:
Use `to-require-context` for glob generation after all

tmeasday committed:
Merge pull request #16008 from storybookjs/on-demand-store-uncached-stories-index

Story index: Uncache `stories.json` generation

shilman committed:
Merge branch 'on-demand-store' into fix/on-demand-store-addon-docs

tmeasday committed:
Fix problem with `prepareForInline`

tmeasday committed:
Update CSF to get rid of play context

shilman committed:
Fix docs rendering

shilman committed:
Merge branch 'on-demand-store' into fix/on-demand-store-addon-docs

tmeasday committed:
Update store/preview to take project annotations on init.

This way they can be set on `window` at boot time.

tmeasday committed:
We need to track initialization now.

shilman committed:
Fix unbalanced hooks error in ArgsTable

tmeasday committed:
Update Store APIs to be backwards-compatible

(Specifically to ensure the links addon works)

shilman committed:
MDX: Rename mdxComponentMeta to mdxComponentAnnotations

shilman committed:
Clean up source snippet retrieval for arbitrary stories

shilman committed:
Fix modern inline rendering with absolutely positioned elements

shilman committed:
Fix stories.json auto-extraction for named exports

tmeasday committed:
Retain hooks between loads of the story store.

shilman committed:
Merge branch 'on-demand-store' into fix/on-demand-store-addon-docs

tmeasday committed:
Ensure we don't call configure for stories if we don't have any

tmeasday committed:
Hooks order fix

shilman committed:
Add simple docs tab e2e test

tmeasday committed:
Wait until all sub-stories have rendered in docs

We need to wait to emit DOCS_RENDERED until all the stories are rendered.

shilman committed:
Update snapshots

shilman committed:
Fix deepscan error

tmeasday committed:
Ensure we emit all `SNIPPET_RENDERED` in `useEffect()`

This also enables us to simplify `SourceContainer`.

shilman committed:
Fix docs test

shilman committed:
Update story to new DocsContext

tmeasday committed:
Fix tests and ensure channel is defined

shilman committed:
Fix failing sourceDecorator tests

shilman committed:
Merge pull request #16037 from storybookjs/on-demand-store-fix-useEffect-sourceContainer

Fix `useEffect` in docs and `SourceContainer`

shilman committed:
Merge branch 'on-demand-store' into fix/on-demand-store-addon-docs

shilman committed:
E2E force remove package-lock.json

shilman committed:
Addon-docs/web-components: Handle no-manifest version check

shilman committed:
Merge pull request #16034 from storybookjs/fix/on-demand-store-addon-docs

Addon-docs: Bugfixes for new on-demand architecture

shilman committed:
Fix deepscan
